orca-queue/src/main/kotlin/com/netflix/spinnaker/orca/q/handler/ContinueParentStageHandler.kt

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright 2017 Netflix, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License")
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.orca.q.handler
+
+import com.netflix.spinnaker.orca.ExecutionStatus.NOT_STARTED
+import com.netflix.spinnaker.orca.ExecutionStatus.SUCCEEDED
+import com.netflix.spinnaker.orca.pipeline.model.Stage
+import com.netflix.spinnaker.orca.pipeline.persistence.ExecutionRepository
+import com.netflix.spinnaker.orca.q.*
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
+import org.springframework.beans.factory.annotation.Autowired
+import org.springframework.beans.factory.annotation.Value
+import org.springframework.stereotype.Component
+import java.time.Duration
+
+@Component
+open class ContinueParentStageHandler
+@Autowired constructor(
+  override val queue: Queue,
+  override val repository: ExecutionRepository,
+  @Value("\${queue.retry.delay.ms:5000}") retryDelayMs: Long
+) : MessageHandler<ContinueParentStage> {
+
+  private val log: Logger = LoggerFactory.getLogger(javaClass)
+  private val retryDelay = Duration.ofMillis(retryDelayMs)
+
+  override fun handle(message: ContinueParentStage) {
+    message.withStage { stage ->
+      if (stage.allBeforeStagesComplete()) {
+        if (stage.hasTasks()) {
+          stage.runFirstTask()
+        } else if (stage.hasAfterStages()) {
+          stage.runAfterStages()
+        } else {
+          queue.push(CompleteStage(stage, SUCCEEDED))
+        }
+      } else {
+        log.warn("Re-queuing $message as other BEFORE stages are still running")
+        queue.push(message, retryDelay)
```

Comment:
If all BEFORE stage branches (or parallel branches) don't appear to be complete it will try again later